pyne/vta.py

```diff
@@ -0,0 +1,481 @@
+    """
+    Move the point to next voxel.
+
+    Parameters:
+    -----------
+    idxs: list
+        List of indexes, format: [x_idx, y_idx, z_idx]
+    tp : numpy array
+        An array of size 3. Temporary point position.
+    t_temp: float
+        Temporary travel lenght from the start.
+    t_maxs: numpy array
+        The maximum travel length in current voxel for 3 directions.
+    v : numpy array
+        Direction vector from start to end.
+
+    Returns:
+    --------
+    idxs: list
+        Updated indexes.
+    tp : numpy array
+        An array of size 3. Point after moving along the direction of v.
+    t_temp : float
+        Updated travel length. Distance from the start point to updated point.
+    """
+    t_min = min(i for i in t_maxs if i > 0)
+    update_dir = list(t_maxs).index(t_min)
+    update_idx = [0, 0, 0]
+    update_idx[update_dir] = 1
+    idxs = [x + y for x, y in zip(idxs, update_idx)] 
+    tp += t_min * v
+    t_temp += t_min
+    return idxs, tp, t_temp
+
+def _move_to_boundary(tp, t_min, v):
+    """
+    Move the point to the nearest boundary when initializing the problem.
+
+    Parameters:
+    -----------
+    tp : numpy array
+        An array of size 3. Temporary point. Actually, it is also the start point.
+    t_min : float
+        Distance to the nearest (positive) boundary. For a 3D mesh, there are 6
+        outside boundaries. Therefore there are 6 value of the distance from the
+        start to the boundaries. This t_min is the smallest one among the
+        positive distance. (The distance is negtive when the direction is
+        negtive.)
+    v : numpy array
+        Direction vector.
+
+    Return:
+    -------
+    tp : numpy array
+        Modified point. This point lies on the nearest boundary. Therefore, we
+        moved point from outside the mesh to the mesh boundary (or the extension
+        of the boundary, which still outside the mesh).
+    """
+
+    tp += t_min * v
+    return tp
+
+def _ray_voxel_traverse(bounds, start, end):
+    """
+    Voxel traversal algorithm.
+    Return the voxels that intersect with the line segment.
+
+    Parameters:
+    -----------
+    bounds : list
+        Boundaries of the mesh grid.
+    start : numpy array
+        Start point.
+    end : numpy array
+        End point.
+
+    Return:
+    -------
+    voxles : list of tuples
+        List of voxel indices. These voxels intersect with the line segment.
+    """
+    # Initialization
+    # Find the voxel that start point in
+    tp = np.copy(start)
+    idxs = [-1, -1, -1]
+    voxels = set()
+    v = _cal_dir_v(start, end)
+    t_end = _distance(start, end)
+    t_temp = 0.0
+    while t_temp < t_end:
+        if _is_point_in_mesh(bounds, tp):
+            # the point is in the mesh
+            # calculate the voxel id
+            for dr in range(len(idxs)):
+                idxs[dr] = _find_voxel_idx_1d(bounds[dr], tp[dr], v[dr])
+            if -1 in idxs:
```

Comment:
Why this extra check? We test if the point is in the mesh above right?

Reply:
The comment here should be revised. 
As the mesh has outside boundaries. A point may be exactly on the boundary. The function `_is_point_in_mesh` will give a `True` value if the point is on the boundary. 
However, we don't need to traverse a point that has a direction toward outside. This is the reason of the extra check here.

Reply:
Then will triangle on the boundary of the mesh not be counted as intersecting with the voxel? Should they be?

I'm feeling like this is a corner case that will be exposed later on...